workspaces/ui-v2/src/store/shapes/thunks.ts

```diff
@@ -0,0 +1,195 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import * as Sentry from '@sentry/react';
+
+import { IForkableSpectacle } from '@useoptic/spectacle';
+import { JsonLike } from '<src>/types';
+import {
+  convertSpectacleChangeToChangeType,
+  SpectacleChange,
+} from '../spectacleUtils';
+import { ShapeId, ReduxShape } from './types';
+
+const ShapeQuery = `
+query X($shapeId: ID! $sinceBatchCommitId: String) {
+  shapeChoices(shapeId: $shapeId) {
+    id
+    jsonType
+    asObject {
+      fields {
+        name
+        fieldId
+        shapeId
+        contributions
+        changes(sinceBatchCommitId: $sinceBatchCommitId) {
+          added
+          changed
+          removed
+        }
+      }
+    }
+    asArray {
+      shapeId
+    }
+  }
+}`;
+
+type SpectacleField = {
+  name: string;
+  fieldId: string;
+  shapeId: string;
+  contributions: Record<string, string>;
+  changes: SpectacleChange;
+};
+
+type SpectacleArray = {
+  shapeId: string;
+};
+
+type SpectacleShape =
+  | {
+      id: string;
+      jsonType: JsonLike.OBJECT;
+      asArray?: undefined;
+      asObject: {
+        fields: SpectacleField[];
+      };
+    }
+  | {
+      id: string;
+      jsonType: JsonLike.ARRAY;
+      asArray: SpectacleArray;
+      asObject?: undefined;
+    }
+  | {
+      id: string;
+      jsonType: Exclude<JsonLike, JsonLike.OBJECT | JsonLike.ARRAY>;
+
+      asArray?: undefined;
+      asObject?: undefined;
+    };
+
+type SpectacleShapeResult = {
+  shapeChoices: SpectacleShape[];
+};
+
+const fetchSpectacleShape = async (
+  spectacle: IForkableSpectacle,
+  shapeId: string,
+  sinceBatchCommitId?: string
+): Promise<SpectacleShapeResult> => {
+  const results = await spectacle.query<
+    SpectacleShapeResult,
+    { shapeId: string; sinceBatchCommitId?: string }
+  >({
+    query: ShapeQuery,
+    variables: { shapeId, sinceBatchCommitId },
+  });
+
+  if (results.errors || !results.data) {
+    console.error(results.errors);
+    throw new Error(JSON.stringify(results.errors));
+  }
+
+  return results.data;
+};
+
+const fetchShapesAndChildren = async (
+  spectacle: IForkableSpectacle,
+  rootShapeId: string,
+  sinceBatchCommitId?: string
+): Promise<Record<ShapeId, ReduxShape[]>> => {
+  // TODO we might want to have access to the current store to
+  // avoid fetching shapes we have already fetched
+  const shapeMap: Record<ShapeId, ReduxShape[]> = {};
+
+  let stack: ShapeId[] = [rootShapeId];
+
+  while (stack.length > 0) {
+    const results = await Promise.all(
+      stack.map((shapeId) =>
+        fetchSpectacleShape(spectacle, shapeId, sinceBatchCommitId)
+      )
+    );
+    // Reset the stack and add in more shapes to fetch
+    stack = [];
```

Comment:
The more standard DFS / BFS would be do visit nodes one at a time, but since this is async, we want to be greedy and query _all_ shapes that we _know_ we have to query.